### ISSUE TITLE
[DependencyInjection] circular references using setter

### ISSUE THREAD

**Issue by ElectricMaxxx** [symfony/symfony#25240]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Symfony version  | 4.0

Hi, 

i think i have an issue with circular references in our Symfony 4.0 build on [DoctrinePhpcrBundle](https://travis-ci.org/doctrine/DoctrinePHPCRBundle/jobs/309869611).

You can reproduce it it by:

```
git clone https://github.com/doctrine/DoctrinePHPCRBundle.git
git checkout allow_symfony_4.0
composer config minimum-stability dev
composer require symfony/symfony:^4.0 --no-update
composer update --prefer-dist
```
(same as travis does it)
Then you can call every console command (lives in testing component, kernel is here):
```
 export KERNEL_CLASS=Doctrine\\Bundle\\PHPCRBundle\\Tests\\Fixtures\\App\\Kernel && ./vendor/symfony-cmf/testing/bin/travis/../console doctrine:phpcr:init:dbal --drop -vvv
```

You should see the issue then. Currently i do no know why? Are we doing something wrong with the [service definition](https://github.com/doctrine/DoctrinePHPCRBundle/blob/master/DependencyInjection/DoctrinePHPCRExtension.php#L500) or is it the setter for the [TranslationStrategy](https://github.com/doctrine/DoctrinePHPCRBundle/blob/master/DependencyInjection/DoctrinePHPCRExtension.php#L520). When debugging into i got the following image at the end where the exception is thrown:

![selection_125](https://user-images.githubusercontent.com/2905834/33477371-bd87850e-d685-11e7-9113-883158867e67.jpg)

Notice:
- The builds on Symfony 3.3, 3.4 and 2.8 are green

**Comment by stof** [comment]
Well, you indeed have a circular reference here: you inject the strategy into the document manager, and the document manager into the strategy. This is very likely to break (writing the code able to instantiate such object graph is quite hard)

**Comment by stof** [comment]
however, having it working on 3.4 and failing on 4.0 looks weird to me.

**Comment by nicolas-grekas** [comment]
The difference must come from some services which have been inlined now that they are really private.
The PhpDumper has some anti-circular checks that might be obsolete now. I'll have a look.

**Comment by ElectricMaxxx** [comment]
Uhhh you are right, did not recoginze that among the code.

**Comment by ElectricMaxxx** [comment]
@stof you got a pattern/hint for me to resolve those references? both classes need each other as it seems. both classes should be created by the DI.

**Comment by nicolas-grekas** [comment]
Fixed in #25247

**Comment by ElectricMaxxx** [comment]
@nicolas-grekas will you release your fix also?

**Comment by nicolas-grekas** [comment]
What does that mean?
The fix will be released with 3.4.1, which should happen in 4-5 weeks max.
